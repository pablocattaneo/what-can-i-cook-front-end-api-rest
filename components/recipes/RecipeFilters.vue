<template>
  <div id="recipe-filters">
    <b-form-checkbox
      :id="`language-${index}`"
      :key="`language-${index}`"
      v-for="(language, index) in languagesData"
      v-model="language.isActive"
      @change="changeLanguage"
      :disabled="language.isDisable"
      v-b-tooltip.hover.top="language.isDisable ? $t('language_disable') : null"
      name="check-button"
      class="d-inline-block mr-4 mt-3"
      switch
    >
      {{ $t(language.text) }}
    </b-form-checkbox>
    <div class="mt-4">
      <label for="range-1">{{ $t('recipes.up_to_calories_by_portion') }}</label>
      <b-form-input
        id="range-1"
        v-model="portionCaloriesData"
        @change="portionCaloriesChange"
        type="range"
        min="0"
        max="1000"
      />
      <div>{{ $t('recipes.portion_calories') }}: {{ portionCaloriesData }}</div>
    </div>
    <WcSelect
      :options="categoryOptions"
      :label="$t('recipes.categories')"
      @input="categoryChange($event)"
      :value="categorySelected"
      class="mt-4"
    />
    <b-button @click="$emit('clear-filter')" class="mt-4">
      <BIconTrash2 />
      {{ $t('clear_filters') }}</b-button
    >
  </div>
</template>

<script>
import WcSelect from '@/components/forms/WcSelect'
import { BIconTrash2 } from 'bootstrap-vue'

export default {
  components: {
    WcSelect,
    BIconTrash2
  },
  props: {
    languages: {
      type: Array,
      default: () => []
    },
    categorySelected: {
      type: String,
      default: ''
    },
    portionCalories: {
      type: Number,
      default: 0
    }
  },
  data() {
    return {
      languagesData: [],
      languagesActivesCounter: 0,
      portionCaloriesData: this.portionCalories,
      categoryOptions: [
        {
          value: null,
          text: this.$t('recipes.categories')
        },
        { value: 'appetizers-and-snacks', text: 'Appetizers & Snacks' },
        { value: 'breakfast-and-brunch', text: 'Breakfast & Brunch' },
        { value: 'desserts', text: 'Desserts' },
        { value: 'dinner', text: 'Dinner' },
        { value: 'drinks', text: 'Drinks' },
        { value: 'lunch', text: 'Lunch' }
      ]
    }
  },
  created() {
    this.languagesData = this.languages
    this.countActivesLanguages()
    this.activeOrDisableLanguageSection()
  },
  methods: {
    countActivesLanguages() {
      this.languages.forEach((language) => {
        if (language.isActive) {
          this.languagesActivesCounter++
        }
      })
    },
    activeOrDisableLanguageSection() {
      if (this.languagesActivesCounter === 1) {
        const activeLangueIndex = this.languages.findIndex(
          (e) => e.isActive === true
        )
        this.languages[activeLangueIndex].isDisable = true
      } else {
        this.languages.forEach((language, index) => {
          this.languages[index].isDisable = false
        })
      }
    },
    changeLanguage() {
      this.$nextTick(() => {
        this.$emit('languages-change', this.languagesData)
      })
    },
    categoryChange(event) {
      if (event) {
        this.$router.push({
          path: this.localePath('/recipes'),
          query: {
            ...this.$route.query,
            category: event
          }
        })
      } else {
        const query = { ...this.$route.query }
        delete query.category
        this.$router.push({
          path: this.localePath('/recipes'),
          query: {
            ...query
          }
        })
      }
    },
    portionCaloriesChange(event) {
      this.$router.push({
        path: this.localePath('/recipes'),
        query: {
          ...this.$route.query,
          calories: event
        }
      })
    }
  }
}
</script>
