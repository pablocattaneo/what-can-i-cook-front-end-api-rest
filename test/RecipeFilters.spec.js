import { mount, createLocalVue } from '@vue/test-utils'
import RecipeFilters from '@/components/recipes/RecipeFilters'

import { BootstrapVue } from 'bootstrap-vue'

const localVue = createLocalVue()
localVue.use(BootstrapVue)

let wrapper
beforeEach(() => {
  wrapper = mount(RecipeFilters, {
    localVue,
    mocks: {
      $t() {}
    },
    propsData: {
      languages: [
        { isActive: true, text: 'es', isDisable: true },
        { isActive: false, text: 'en', isDisable: false }
      ]
    }
  })
})

test('Sanity test', () => {
  expect(true).toBe(true)
})

describe('Snapshots', () => {
  test('Branch: default. It verify if non-intentional changes were made.', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})

describe('Props', () => {
  test('The component is well formed so prop languages should exist', () => {
    expect(wrapper.props()).toHaveProperty('languages')
  })
  test('The props languages was set to [{"isActive":true,"text":"es","isDisable":true},{"isActive":false,"text":"en","isDisable":false}] so #language-0 and #language-1 elements should be render.', async () => {
    await wrapper.setProps({
      languages: [
        { isActive: true, text: 'es', isDisable: true },
        { isActive: false, text: 'en', isDisable: false }
      ]
    })
    expect(wrapper.findAll('#language-0, #language-1')).toHaveLength(2)
  })
  test('The component is well formed so prop categorySelected should exist', () => {
    expect(wrapper.props()).toHaveProperty('categorySelected')
  })
  test('The component is well formed so prop portionCalories should exist', () => {
    expect(wrapper.props()).toHaveProperty('portionCalories')
  })
})
