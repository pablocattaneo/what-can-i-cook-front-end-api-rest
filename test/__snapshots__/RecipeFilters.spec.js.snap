// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots Branch: default. It verify if non-intentional changes were made. 1`] = `
<div
  id="recipe-filters"
>
  <div
    class="d-inline-block mr-4 mt-3 custom-control custom-switch"
  >
    <input
      class="custom-control-input"
      disabled="disabled"
      id="language-0"
      name="check-button"
      type="checkbox"
      value="true"
    />
    <label
      class="custom-control-label"
      for="language-0"
    >
      
    
  
    </label>
  </div>
  <div
    class="d-inline-block mr-4 mt-3 custom-control custom-switch"
  >
    <input
      class="custom-control-input"
      id="language-1"
      name="check-button"
      type="checkbox"
      value="true"
    />
    <label
      class="custom-control-label"
      for="language-1"
    >
      
    
  
    </label>
  </div>
   
  <div
    class="mt-4"
  >
    <label
      for="range-1"
    >
      
    </label>
     
    <input
      class="custom-range"
      id="range-1"
      max="1000"
      min="0"
      type="range"
    />
     
    <div>
      : 0
    </div>
  </div>
   
  <fieldset
    class="form-group mt-4"
    for=""
    id="label-"
  >
    <legend
      class="bv-no-focus-ring col-form-label pt-0"
      id="label-__BV_label_"
      tabindex="-1"
    >
      
    
    
      <!---->
    </legend>
    <div>
       
      <select
        class="custom-select"
        id="__BVID__21"
      >
        <option
          value=""
        >
          null
        </option>
        <option
          value="appetizers-and-snacks"
        >
          Appetizers & Snacks
        </option>
        <option
          value="breakfast-and-brunch"
        >
          Breakfast & Brunch
        </option>
        <option
          value="desserts"
        >
          Desserts
        </option>
        <option
          value="dinner"
        >
          Dinner
        </option>
        <option
          value="drinks"
        >
          Drinks
        </option>
        <option
          value="lunch"
        >
          Lunch
        </option>
      </select>
       
      <div
        class="invalid-feedback"
      >
        
    
  
      </div>
       
      <!---->
      <!---->
      <!---->
      <!---->
    </div>
  </fieldset>
   
  <button
    class="btn mt-4 btn-secondary"
    type="button"
  >
    <svg
      aria-label="trash2"
      class="bi-trash2 b-icon bi"
      fill="currentColor"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g>
        <path
          d="M14 3a.702.702 0 0 1-.037.225l-1.684 10.104A2 2 0 0 1 10.305 15H5.694a2 2 0 0 1-1.973-1.671L2.037 3.225A.703.703 0 0 1 2 3c0-1.105 2.686-2 6-2s6 .895 6 2zM3.215 4.207l1.493 8.957a1 1 0 0 0 .986.836h4.612a1 1 0 0 0 .986-.836l1.493-8.957C11.69 4.689 9.954 5 8 5c-1.954 0-3.69-.311-4.785-.793z"
        />
      </g>
    </svg>
    
    
  </button>
</div>
`;
